lib/entice/logic/MapInstance.ex

```diff
@@ -0,0 +1,67 @@
+defmodule Entice.Logic.MapInstance do
+  alias Entice.Entity
+  alias Entice.Logic.MapInstance
+
+  defstruct(
+    players: [],
+    npcs: [],
+    map: nil)
+
+  def register(entity, map),
+  do: Entity.put_behaviour(entity, MapInstance.Behaviour, map)
+
+  def unregister(entity),
+  do: Entity.remove_behaviour(entity, MapInstance.Behaviour)
+
+  defmodule Behaviour do
+    use Entice.Entity.Behaviour
+
+    def init(entity, map),
+    do: {:ok, entity |> put_attribute(%MapInstance{map: map})}
+
+    def handle_call({:map_instance_player_join, player}, entity) when is_pid(player) do
+      add_creature_to_map_instance(entity, player, :players)
+    end
+
+    def handle_call({:map_instance_npc_join, npc}, entity) when is_pid(npc) do
+      add_creature_to_map_instance(entity, npc, :npcs)
+    end
+
+    def handle_call({:map_instance_player_leave, player}, entity) when is_pid(player) do
+      {:ok, entity, empty} = remove_creature_from_map_instance(entity, player, :players)
+      case empty do
+        false -> {:ok, entity}
+        _ -> {:stop, :normal, entity}
+      end
+    end
+
+    def handle_call({:map_instance_npc_leave, npc}, entity) when is_pid(npc) do
+      {:ok, entity, _} = remove_creature_from_map_instance(entity, npc, :npcs)
+      {:ok, entity}
+    end
+
+    def handle_call(event, entity), do: super(event, entity)
+
+    def terminate(_reason, entity) do
+      {:ok, entity |> remove_attribute(MapInstance)}
+    end
+
+    defp add_creature_to_map_instance(%Entity{attributes: %{MapInstance => map_instance}} = entity, creature, creature_key) do
```

Comment:
i think we dont need these two helpers if we take the counter :)
but for the future: you can also use the `Set` module, and `HashSet`, `MapSet` etc., see the elixir docs